### PR TITLE
Improve training performance

### PR DESCRIPTION
**Pull request by aribrill**
This PR makes training models with CTLearn faster and more correct.

First, the input function has been restructured to use `Dataset.from_tensor_slices()` and `Dataset.map()` for simplicity and performance. With https://github.com/cta-observatory/dl1-data-handler/pull/88, using the `num_parallel_calls` argument of `Dataset.map()` works, enabling a ~40% speedup on my machine. In addition, `tf.data.experimental.prefetch_to_device()` is used to prefetch examples directly to GPU, providing another 20-25% speedup.

Second, the training and validation loop has been restructured to properly shuffle and use all of the data. Previously, only a small buffer of examples was shuffled and trained on during each iteration. In particular, when a random seed was set, the shuffling would be deterministic on each iteration, so the same set of examples would be used each time. As a result, only a small subset of the training set was actually used for training!

This PR fixes this problem by using `tf.estimator.train_and_evaluate()` to construct the training input function only once, combined with `Dataset.repeat()` to allow training for an indefinite number of epochs. Because of the previous changes to the input function, only the example identifiers are shuffled, allowing for a perfect shuffle for virtually any sized dataset. To support this API, the max_steps argument is used to determine the total number of steps to train, as opposed to the number of validations and steps between validations.

Finally, the now-obsolete configuration parameters `shuffle_buffer_size`, `prefetch_buffer_size`, `num_validations`, and `num_training_steps_per_validation` have been removed, and new parameters `prefetch_to_device` and `max_steps` have been added to support the new features.

### CODE REVIEW

File: config/example_config.yml [hunk 1]
```diff
@@ -404,14 +402,9 @@ Training:
...
-    # Number of validations made before finishing training. If 0, run forever.
-    num_validations: 0
-
-    # Required integer.
-    # Number of training steps to run before each evaluation
-    # on the validation set.
-    num_training_steps_per_validation: 1000
+    # Required int.
+    # Number of steps (usually, batches) on which to train.
+    max_steps: 10000
```
Comment (TjarkMiener) [inline config/example_config.yml:407]: What do you think of getting the max_steps values from num_epoch * training_steps_per_epoch? I think it's more intuitive and accepted by the DL community to show our metrics vs epochs.

File: ctlearn/run_model.py [hunk 2]
```diff
@@ -254,23 +252,25 @@ def run_model(config, mode="train", debug=False, log_to_file=False):
...
 
-        def generator(indices):
-            for idx in indices:
-                yield tuple(reader[idx])
+        dataset = tf.data.Dataset.from_tensor_slices(indices)
+        if shuffle_and_repeat:
+            dataset = dataset.shuffle(buffer_size=len(indices), seed=seed,
+                                      reshuffle_each_iteration=True)
+            dataset = dataset.repeat()
+        dataset = dataset.map(lambda x: tf.py_function(func=reader.__getitem__,
```
Comment (TjarkMiener) [inline ctlearn/run_model.py:265]: This is very nice, because we can add a function in the reader to return the output in the shape `(features, labels)`. That would be needed for the TF2 update. Is this true?

File: ctlearn/run_model.py [hunk 3]
```diff
@@ -439,19 +439,16 @@ def model_fn(features, labels, mode, params):
...
-                num_validations_remaining -= 1
+        max_steps = config['Training']['max_steps']
+        train_input_fn = lambda: input_fn(reader, training_indices,
+                                          shuffle_and_repeat=True,
+                                          **config['Input'])
+        train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn,
+                                            max_steps=max_steps, hooks=hooks)
+        eval_input_fn = lambda: input_fn(reader, validation_indices,
+                                         **config['Input'])
+        eval_spec = tf.estimator.EvalSpec(input_fn=eval_input_fn)
```
Comment (TjarkMiener) [inline ctlearn/run_model.py:450]: I would be in favour of passing the number of training steps per epoch for the `steps` argument, so we evaluate after each epoch. What do you think?

File: config/example_config.yml (hunk 1 above)
Comment (aribrill) [reply config/example_config.yml:407]: Sure, that makes sense to me.

File: ctlearn/run_model.py (hunk 2 above)
Comment (aribrill) [reply ctlearn/run_model.py:265]: The output is already returned in this format (see line 288), unless I missed something?

File: ctlearn/run_model.py (hunk 3 above)
Comment (aribrill) [reply ctlearn/run_model.py:450]: The `steps` argument in the `EvalSpec` refers to the number of validation steps, not the number of steps between validations. We want the default (evaluate on the whole validation set). In the `train_and_evaluate` API, the separation between evaluations is based on time, not number of steps. It will evaluate every 2 minutes by default (skipping if there's no new checkpoint). The time between validations is set by the `start_delay_secs` argument. We could make that configurable, though I'm not sure there's a strong need for it.

File: ctlearn/run_model.py (hunk 2 above)
Comment (TjarkMiener) [reply ctlearn/run_model.py:265]: Sorry my comment wasn't very precise! With `(features, labels)` I meant a 2D tuple of tuples. The current TF2 input, which is returning a dataset looks like this:
```
def input_fn(reader, indices, output_dtypes, output_shapes,
             feature_positions, label_positions, mode='train', seed=None, batch_size=1,
             shuffle_buffer_size=None, prefetch_buffer_size=1,
             add_labels_to_features=False):
    
    def generator(indices):
        for idx in indices:
            yield (map(lambda fpos: reader[idx][fpos], feature_positions), map(lambda lpos: reader[idx][lpos], label_positions))
    
    dataset = tf.data.Dataset.from_generator(generator, output_dtypes,
                                             output_shapes=output_shapes,
                                             args=(indices,))
   
    # Only shuffle the data, when train mode is selected.
    if mode == 'train':
        if shuffle_buffer_size is None:
            shuffle_buffer_size = len(indices)
        dataset = dataset.shuffle(buffer_size=shuffle_buffer_size, seed=seed)
    dataset = dataset.batch(batch_size)
    dataset = dataset.prefetch(prefetch_buffer_size)
    
    return dataset
```
 What we can do is to add a function`reader.__getitem_TF2__` with arguments `feature_positions` and `label_positions` and splitting the reader output in the right format. Then we can return the the dataset from the `input_fn()` and skip the custom separation from line 275 - 286.

File: ctlearn/run_model.py (hunk 3 above)
Comment (TjarkMiener) [reply ctlearn/run_model.py:451]: Agree!

File: config/example_config.yml (hunk 1 above)
Comment (TjarkMiener) [reply config/example_config.yml:406]: Good! The argument should be also renamed to `num_epochs` here, since we are using `config['Training']['num_epochs']` in `run_model.py` line 233.
